File: workingtime/src/main/java/io/nirahtech/erp/workingtime/Imputable.java
package io.nirahtech.erp.workingtime;

import java.time.LocalDate;
import java.util.Collection;
import java.util.concurrent.TimeUnit;

public interface Imputable {
    void impute(final Imputation workingTime);
    int computeWorkedTime(TimeUnit timeUnit);
    int computeWorkedTime(LocalDate from, LocalDate to, TimeUnit timeUnit);
    
    int computeWorkedTimeOnProject(TimeUnit timeUnit, Project project);
    int computeWorkedTimeOnProject(LocalDate from, LocalDate to, TimeUnit timeUnit, Project project);

    Collection<Imputation> listImputations();
}

File: webapp/src/main/react/src/data/Imputable.ts
import Imputation from './Imputation';
import LocalDateTime from './LocalDateTime';
import Project from './Project';
import TimeUnit from './TimeUnit';

interface Imputable {
    impute(workingTime: Imputation): void;
    computeWorkedTimeUsingTimeUnit(timeUnit: TimeUnit): number;
    computeWorkedTimeFromDatesRangeUsingTimeUnit(from: LocalDateTime | null, to: LocalDateTime | null, timeUnit: TimeUnit): number;
    computeWorkedTimeOnProjectUsingTimeUnit(timeUnit: TimeUnit, project: Project | null): number;
    computeWorkedTimeOnProject(
        from: LocalDateTime | null,
        to: LocalDateTime | null,
        timeUnit: TimeUnit,
        project: Project | null
    ): number;
    listImputations(): ReadonlyArray<Imputation>;
}

export default Imputable